// ernest-logger/index.js
const fs = require('fs');
const path = require('path');
const colors = require('./colors/colors.js');
const emojis = require('./emojis/emojis.js');
const getTimestamp = require('./utils/timestamp.js');
const formatter = require('./utils/formatter.js');

// Priority map for filtering logs
const levelPriority = {
    trace: 0,
    debug: 1,
    info: 2,
    success: 3,
    warn: 4,
    error: 5,
    fatal: 6,
};

// Smart default configuration
let globalConfig = {
    time: true,
    emoji: true,
    level: 'debug',
    file: false,
    filePath: path.join(process.cwd(), 'ernest.log'),
    colorize: true,
    prefix: '',
    maxFileSize: 5 * 1024 * 1024, // 5MB
    customLevels: {},
};

// File rotation logic
const rotateLogFile = (filePath, maxSize) => {
    try {
        const stats = fs.statSync(filePath);
        if (stats.size > maxSize) {
            const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
            const rotatedPath = filePath.replace('.log', `-${timestamp}.log`);
            fs.renameSync(filePath, rotatedPath);
        }
    } catch (err) {
        // File doesn't exist yet, no rotation needed
    }
};

const createLogger = (options = {}) => {
    const config = { ...globalConfig, ...options };

    const formatMessage = (level, message, color) => {
        if (levelPriority[level] < levelPriority[config.level]) return null;

        let output = '';
        
        // Add timestamp if enabled
        if (config.time) {
            const timestamp = getTimestamp();
            output += config.colorize ? colors.dim + `[${timestamp}]` + colors.reset + ' ' : `[${timestamp}] `;
        }

        // Add prefix if configured
        if (config.prefix) {
            output += config.colorize ? colors.dim + config.prefix + colors.reset + ' ' : config.prefix + ' ';
        }

        // Add emoji if enabled
        if (config.emoji && emojis[level]) {
            output += `${emojis[level]} `;
        }

        // Add level badge with color
        const levelBadge = formatter.createBadge(level.toUpperCase(), color);
        output += config.colorize ? levelBadge + ' ' : `[${level.toUpperCase()}] `;

        // Add the actual message with color
        output += config.colorize ? colors[color] + message + colors.reset : message;

        return output;
    };

    const writeToFile = (message) => {
        if (!config.file) return;
        
        try {
            rotateLogFile(config.filePath, config.maxFileSize);
            const plainMessage = message.replace(/\x1b\[[0-9;]*m/g, ''); // Strip ANSI codes
            fs.appendFileSync(config.filePath, plainMessage + '\n');
        } catch (err) {
            console.error('Failed to write to log file:', err.message);
        }
    };

    const logMessage = (level, color, message) => {
        const formatted = formatMessage(level, message, color);
        if (!formatted) return;

        console.log(formatted);
        writeToFile(formatted);
    };

    // Create the main logger object with smart defaults
    const logger = {
        // Trace level (most verbose)
        trace: (msg) => logMessage('trace', 'dim', msg),

        // Debug level
        debug: (msg) => logMessage('debug', 'cyan', msg),

        // Info level (default)
        info: (msg) => logMessage('info', 'blue', msg),

        // Success level
        success: (msg) => logMessage('success', 'green', msg),

        // Warn level
        warn: (msg) => logMessage('warn', 'yellow', msg),

        // Error level
        error: (msg) => logMessage('error', 'red', msg),

        // Fatal level (most critical)
        fatal: (msg) => logMessage('fatal', 'brightRed', msg),

        // Specialized quick methods
        start: (msg) => logMessage('info', 'brightGreen', `▶️  ${msg}`),
        stop: (msg) => logMessage('info', 'brightRed', `⏹️  ${msg}`),
        network: (msg) => logMessage('info', 'brightCyan', `🌐 ${msg}`),
        db: (msg) => logMessage('info', 'brightMagenta', `💾 ${msg}`),
        api: (msg) => logMessage('info', 'brightBlue', `🔌 ${msg}`),
        security: (msg) => logMessage('warn', 'brightYellow', `🔒 ${msg}`),
        
        // Big log for announcements
        bigLog: (message, options = {}) => {
            const color = options.color || 'blue';
            const emoji = options.emoji || '📣';
            const boxed = formatter.createBox(message, color, emoji);
            console.log(boxed);
            if (config.file) writeToFile(boxed);
        },

        // Table logger
        table: (data) => {
            console.table(data);
            if (config.file) {
                writeToFile('\n[TABLE DATA]\n' + JSON.stringify(data, null, 2));
            }
        },

        // Group logs
        group: (label) => console.group(label),
        groupEnd: () => console.groupEnd(),

        // Timer utilities
        time: (label) => console.time(label),
        timeEnd: (label) => console.timeEnd(label),

        // JSON pretty print
        json: (obj, label = '') => {
            const output = label ? `${label}:\n` : '';
            const formatted = output + JSON.stringify(obj, null, 2);
            console.log(config.colorize ? colors.cyan + formatted + colors.reset : formatted);
            if (config.file) writeToFile(formatted);
        },

        // Advanced nested logger (color-coded by level)
        log: {
            trace: {
                dim: (msg) => logMessage('trace', 'dim', msg),
                white: (msg) => logMessage('trace', 'white', msg),
            },
            debug: {
                cyan: (msg) => logMessage('debug', 'cyan', msg),
                blue: (msg) => logMessage('debug', 'blue', msg),
                magenta: (msg) => logMessage('debug', 'magenta', msg),
            },
            info: {
                blue: (msg) => logMessage('info', 'blue', msg),
                cyan: (msg) => logMessage('info', 'cyan', msg),
                white: (msg) => logMessage('info', 'white', msg),
            },
            success: {
                green: (msg) => logMessage('success', 'green', msg),
                brightGreen: (msg) => logMessage('success', 'brightGreen', msg),
            },
            warn: {
                yellow: (msg) => logMessage('warn', 'yellow', msg),
                brightYellow: (msg) => logMessage('warn', 'brightYellow', msg),
            },
            error: {
                red: (msg) => logMessage('error', 'red', msg),
                brightRed: (msg) => logMessage('error', 'brightRed', msg),
                magenta: (msg) => logMessage('error', 'magenta', msg),
            },
            fatal: {
                brightRed: (msg) => logMessage('fatal', 'brightRed', msg),
                bgRed: (msg) => logMessage('fatal', 'bgRed', msg),
            },
        },

        // Configuration methods
        configure: (newConfig) => {
            Object.assign(config, newConfig);
            return logger;
        },

        getConfig: () => ({ ...config }),

        // Enable/disable features on the fly
        enableTime: () => { config.time = true; return logger; },
        disableTime: () => { config.time = false; return logger; },
        enableEmoji: () => { config.emoji = true; return logger; },
        disableEmoji: () => { config.emoji = false; return logger; },
        enableFile: () => { config.file = true; return logger; },
        disableFile: () => { config.file = false; return logger; },
        setLevel: (level) => { config.level = level; return logger; },
    };

    // Inject custom levels dynamically
    if (config.customLevels && Object.keys(config.customLevels).length > 0) {
        Object.entries(config.customLevels).forEach(([level, { color, emoji, priority }]) => {
            if (emoji) emojis[level] = emoji;
            if (priority !== undefined) levelPriority[level] = priority;
            
            logger[level] = (msg) => logMessage(level, color, msg);
            logger.log[level] = {
                [color]: (msg) => logMessage(level, color, msg),
            };
        });
    }

    return logger;
};

// Create a default logger instance that's ready to use
const defaultLogger = createLogger();

// Export both the factory and the default instance
module.exports = defaultLogger;
module.exports.createLogger = createLogger;
module.exports.setGlobalConfig = (config) => {
    Object.assign(globalConfig, config);
};
