// ernest-logger/colors/colors.js
/**
 * Enhanced ANSI escape codes for console text colors and styles.
 * Supports 16 colors, bright variants, background colors, and text styling.
 */

const colors = {
  // Standard colors
  black: "\x1b[30m",
  red: "\x1b[31m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  blue: "\x1b[34m",
  magenta: "\x1b[35m",
  cyan: "\x1b[36m",
  white: "\x1b[37m",

  // Bright/Bold colors
  brightBlack: "\x1b[90m",
  brightRed: "\x1b[91m",
  brightGreen: "\x1b[92m",
  brightYellow: "\x1b[93m",
  brightBlue: "\x1b[94m",
  brightMagenta: "\x1b[95m",
  brightCyan: "\x1b[96m",
  brightWhite: "\x1b[97m",

  // Background colors
  bgBlack: "\x1b[40m",
  bgRed: "\x1b[41m",
  bgGreen: "\x1b[42m",
  bgYellow: "\x1b[43m",
  bgBlue: "\x1b[44m",
  bgMagenta: "\x1b[45m",
  bgCyan: "\x1b[46m",
  bgWhite: "\x1b[47m",

  // Bright background colors
  bgBrightBlack: "\x1b[100m",
  bgBrightRed: "\x1b[101m",
  bgBrightGreen: "\x1b[102m",
  bgBrightYellow: "\x1b[103m",
  bgBrightBlue: "\x1b[104m",
  bgBrightMagenta: "\x1b[105m",
  bgBrightCyan: "\x1b[106m",
  bgBrightWhite: "\x1b[107m",

  // Text styles
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  italic: "\x1b[3m",
  underline: "\x1b[4m",
  blink: "\x1b[5m",
  reverse: "\x1b[7m",
  hidden: "\x1b[8m",
  strikethrough: "\x1b[9m",

  // Semantic aliases for better readability
  success: "\x1b[32m",      // green
  error: "\x1b[31m",        // red
  warning: "\x1b[33m",      // yellow
  info: "\x1b[34m",         // blue
  debug: "\x1b[36m",        // cyan
  
  // Additional semantic colors
  primary: "\x1b[94m",      // brightBlue
  secondary: "\x1b[90m",    // brightBlack (gray)
  danger: "\x1b[91m",       // brightRed
  muted: "\x1b[2m",         // dim
};

/**
 * Utility function to combine multiple color codes
 * @param {...string} codes - Color code names
 * @returns {string} Combined ANSI codes
 */
colors.combine = (...codes) => {
  return codes.map(code => colors[code] || '').join('');
};

/**
 * Utility to wrap text in a color with auto-reset
 * @param {string} text - Text to colorize
 * @param {string} color - Color name
 * @returns {string} Colorized text
 */
colors.wrap = (text, color) => {
  return (colors[color] || '') + text + colors.reset;
};

/**
 * Create a gradient effect (simple simulation)
 * @param {string} text - Text to apply gradient
 * @param {string[]} colorSequence - Array of color names
 * @returns {string} Text with gradient colors
 */
colors.gradient = (text, colorSequence = ['cyan', 'blue', 'magenta']) => {
  const chars = text.split('');
  const colorCount = colorSequence.length;
  
  return chars.map((char, i) => {
    const colorName = colorSequence[i % colorCount];
    return colors[colorName] + char;
  }).join('') + colors.reset;
};

/**
 * Rainbow effect
 * @param {string} text - Text to rainbowify
 * @returns {string} Rainbow colored text
 */
colors.rainbow = (text) => {
  const rainbowColors = ['red', 'yellow', 'green', 'cyan', 'blue', 'magenta'];
  return colors.gradient(text, rainbowColors);
};

module.exports = colors;
