// ernest-logger/utils/timestamp.js
/**
 * Enhanced timestamp utilities with multiple formats.
 */

/**
 * Returns a formatted timestamp string in HH:MM:SS format.
 * @returns {string} The timestamp in 'HH:MM:SS' format.
 */
const getTimestamp = () => {
    const date = new Date();
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    const seconds = String(date.getSeconds()).padStart(2, '0');
    return `${hours}:${minutes}:${seconds}`;
};

/**
 * Returns timestamp with milliseconds.
 * @returns {string} Timestamp in 'HH:MM:SS.mmm' format.
 */
const getTimestampWithMs = () => {
    const date = new Date();
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    const seconds = String(date.getSeconds()).padStart(2, '0');
    const ms = String(date.getMilliseconds()).padStart(3, '0');
    return `${hours}:${minutes}:${seconds}.${ms}`;
};

/**
 * Returns full date and time.
 * @returns {string} Timestamp in 'YYYY-MM-DD HH:MM:SS' format.
 */
const getFullTimestamp = () => {
    const date = new Date();
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    const seconds = String(date.getSeconds()).padStart(2, '0');
    return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
};

/**
 * Returns ISO 8601 timestamp.
 * @returns {string} ISO formatted timestamp.
 */
const getISOTimestamp = () => {
    return new Date().toISOString();
};

/**
 * Returns Unix timestamp in seconds.
 * @returns {number} Unix timestamp.
 */
const getUnixTimestamp = () => {
    return Math.floor(Date.now() / 1000);
};

/**
 * Returns relative time string (e.g., "2 seconds ago").
 * @param {Date|number} date - Date object or timestamp.
 * @returns {string} Relative time string.
 */
const getRelativeTime = (date) => {
    const now = Date.now();
    const timestamp = date instanceof Date ? date.getTime() : date;
    const diff = now - timestamp;
    
    const seconds = Math.floor(diff / 1000);
    const minutes = Math.floor(seconds / 60);
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);
    
    if (seconds < 60) return `${seconds} second${seconds !== 1 ? 's' : ''} ago`;
    if (minutes < 60) return `${minutes} minute${minutes !== 1 ? 's' : ''} ago`;
    if (hours < 24) return `${hours} hour${hours !== 1 ? 's' : ''} ago`;
    return `${days} day${days !== 1 ? 's' : ''} ago`;
};

/**
 * Format timestamp with custom format.
 * Supported tokens: YYYY, MM, DD, HH, mm, ss, SSS
 * @param {string} format - Format string.
 * @param {Date} date - Date object (defaults to now).
 * @returns {string} Formatted timestamp.
 */
const formatCustom = (format, date = new Date()) => {
    const tokens = {
        YYYY: date.getFullYear(),
        MM: String(date.getMonth() + 1).padStart(2, '0'),
        DD: String(date.getDate()).padStart(2, '0'),
        HH: String(date.getHours()).padStart(2, '0'),
        mm: String(date.getMinutes()).padStart(2, '0'),
        ss: String(date.getSeconds()).padStart(2, '0'),
        SSS: String(date.getMilliseconds()).padStart(3, '0'),
    };
    
    let result = format;
    Object.entries(tokens).forEach(([token, value]) => {
        result = result.replace(token, value);
    });
    
    return result;
};

module.exports = getTimestamp;
module.exports.getTimestamp = getTimestamp;
module.exports.getTimestampWithMs = getTimestampWithMs;
module.exports.getFullTimestamp = getFullTimestamp;
module.exports.getISOTimestamp = getISOTimestamp;
module.exports.getUnixTimestamp = getUnixTimestamp;
module.exports.getRelativeTime = getRelativeTime;
module.exports.formatCustom = formatCustom;
